keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +337,302 @@ or later. The dual-stack feature requires no changes to this API.
 The versions of CNI plugin binaries that must be used for proper dual-stack functionality (and IPv6 functionality in general) depend upon the version of the container runtime that is used in the cluster nodes (see [CNI issue #531](https://github.com/containernetworking/cni/issues/531) and [CNI plugins PR #113](https://github.com/containernetworking/plugins/pull/113)):
 
 ### Services
+Services depend on `--service-cluster-ip-range` for `ClusterIP` assignment. A dualstack cluster can have this flag configured as `--services-cluster-ip-rang=<cidr>,<cidr>` or `--services-cluster-ip-rang=<cidr>`. The following rules apply:
+
+1. The maximum allowed number of cidrs in this flag is two. Attempting to use more than two cidrs will generate failures during startup. 
+2. When configured with two cidrs then the values are expected to have different ip families. e.g `--service-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>` or `--service-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`. Attempting to configure a cluster with two similar cidrs will generate failures during startup.
+3. It is possible to have a `partially-dual-stack` cluster, with dual-stack pod networking but only a single service cidr. This means that pods in the cluster will be able to send and receive traffic on both ip families, but services which use the `clusterIP` or `ClusterIPs` field will only be able to use the ip family of this flag.
+4. When upgrading a cluster to dualstack, the first element of the `--service-cluster-ip-range` flag MUST NOT change IP families.  As an example, an IPv4 cluster can safely be upgraded to use `--service-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>`, but can not safely be changed to `--service-cluster-ip-range=<ipv6-cidr>,<ipv4-cidr>`.
+5. The first cidr entry of `--service-cluster-ip-range=<first cidr entry>, <secondary cidr entry>` is considered cluster default ip address family.
+6. Attempting to allocate `ClusterIP` or `ClusterIPs` from an ip address family that is not configured on the cluster will result in an error.
+
+> The below discussion assumes that `Endpoint Slice and Endpoint Slice Controller`  will allow dual stack endpoints, the existing endpoint controller will remain single stack. That means enabling dual stack services by having more than one entry in `--service-cluster-ip-range` without enabling `Endpoint Slice` feature gate will result in validation failures during startup.
+
+#### Kubernetes Default Service
+
+> discussion of `spec.preferDualStack` and `spec.ipFamilies` fields is detailed in the next section.
+
+When the api-server is configured with more than one cidr in `--service-cluster-ip-range` flag. api-server will automatically recreate dualstack enabled kubernetes default service. The service will carry `.spec.ClusterIP` of the same ip family as cluster default ip address family. The rest of the fields will be set according to the same service below for a cluster configured with `--service-cluster-ip-range=<ipv4-cidr>,<ipv6-cidr>`
```

Comment:
Good to cover this, but I one point.  During the upgrade, some API servers will be dual-stack and some will not.  Will they fight over this?  Maybe not, but I am not sure.  The more I think on it, it will PROBABLY be OK, but we should audit this to be sure

Reply:
I have to check the code. the controller that lives inside api-server has no lease. as in they already fighting. I can add a lease and force one controller at a time.

Reply:
It appears that if the Service already exists, 1.18 apiservers will never try to modify the ClusterIP. So if we implement only what it says here ("apiserver will autmoatically convert _to_ dual-stack") then we're fine; the first dual-stack apiserver to come up will change the Service, and no one else will try to change it back.

If we also wanted to implement "apiserver will automatically downgrade to single-stack if you revert the dual-stack configuration", _then_ there would be problems, since during the migration, a single-stack 1.19 apiserver and a dual-stack 1.19 apiserver would each believe they needed to fix the config that the other one set.

On the endpoint side, for `Endpoints` there will also not be problems because apiservers only add and remove themselves from the `Endpoints` object and don't touch other IPs there (and we need to be single-stack there anyway).

For `EndpointSlice`, there may be problems, but this depends on how EndpointSlices work now, which the KEP doesn't specify. But 1.18 apiservers assume there will be a single `EndpointSlice` which has the same name as the `Service` and a single-stack `addressType`, and they will fix it to have that if it's wrong.

Reply:
it does not look there is a lease wrapping `managing kubernetes default service`. api-server primarily relay get->check(if not created goto)->patch->(fail? yes: start over). I am open to think and have it as part of this PR.

While - if my findings are correct - it is suboptimal, I do not believe that will have any impact on `dualstacking kubernetes default service`. Because current code checks if service exists and have proper ip, while the modified code will check for ip families.

Reply:
I see tow results from this thread:

1) Decide/document - is it possible to move from dual-stack back to single-stack?
2) EndpointSlice support for kubernetes.default

Reply:
we removed dualstack for kubernetes default.